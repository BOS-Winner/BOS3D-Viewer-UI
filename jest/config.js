module.exports = {
  rootDir: '../',
  testMatch: [
    "**/__tests__/**/*.test.{js,jsx}",
  ],
  moduleNameMapper: {
    "\\.less$": "identity-obj-proxy",
    "\\.css$": "<rootDir>/jest/__mocks__/styleMock.js",
    "Libs(.*)$": "<rootDir>/src/Libs$1",
    "UIutils(.*)$": "<rootDir>/src/UI/utils$1",
    "^Base(.*)$": "<rootDir>/src/UI/Base$1",
    "^BOS3D(.*)$": "<rootDir>/jest/__mocks__/BOS3D$1",
    "^BOS2D$": "<rootDir>/jest/__mocks__/BOS2D.js",
    "^BOS3DUI$": "<rootDir>/jest/__mocks__/BOS3DUI.js",
    "^jest/(.*)$": "<rootDir>/jest/$1",
    "^mock/(.*)$": "<rootDir>/jest/__mocks__/$1",
  },
  moduleFileExtensions: ["js", "jsx"],
  moduleDirectories: ["node_modules"],
  transform: {
    "\\.(js|jsx)": "babel-jest",
    "\\.(jpg|jpeg|png|gif|svg)$": "<rootDir>/jest/__mocks__/fileMock.js",
  },
  transformIgnorePatterns: [
    // Change MODULE_NAME_HERE to your module that isn't being compiled
    // "/node_modules/(?!three|lodash-es).+\\.js$"
    "/node_modules/core-js",
    "/node_modules/@testing-library",
    "/node_modules/@babel",
  ],
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/jest/"
  ],
};
