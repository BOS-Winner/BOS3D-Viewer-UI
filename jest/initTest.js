import EventEmitter from "events";
import { createStore } from "redux";
import BOS3D from 'BOS3D';
import reducer from "../src/UI/reducer";

export function genStore() {
  const ee = new EventEmitter();
  const viewer3D = new BOS3D.Viewer();

  return createStore(reducer, {
    system: {
      viewer3D,
      BIMWINNER: { BOS3D },
      apiVersion: 'api',
      eventEmitter: ee,
    }
  });
}
