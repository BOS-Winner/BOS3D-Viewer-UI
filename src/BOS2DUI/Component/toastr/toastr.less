:global{
  #toast-container{
    position: absolute;
    .toast{
      border-radius: 4px;
      opacity:1;
      padding: 10px 10px 10px 40px;
    }

    >.toast-error{
      color: #333333;
      border: 1px solid #F03D3D;
      background-color: #FFF1F1;
      background-image:url('./img/error.png')!important;
    }
    >.toast-success{
      color: #333333;
      border: 1px solid #3ED3A3;
      background-color: #E8FFEB;
      background-image:url('./img/success.png')!important;
    }
    >.toast-info{
      color: #333333;
      border: 1px solid #2878FF;
      background-color: #FFFFFF;
      background-image:url('./img/info.png')!important;
    }
    >.toast-warning{
      color: #333333;
      border: 1px solid #FFAD0D;
      background-color: #FFFAEB;
      background-image:url('./img/warning.png')!important;
    }
  }
}

