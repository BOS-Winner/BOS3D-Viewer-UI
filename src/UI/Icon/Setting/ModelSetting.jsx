import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import _ from "lodash-es";
import { Select } from 'antd';
import { changeModelSetting } from "../../userRedux/userSetting/action";
import toastr from "../../toastr";
import style from "./style.less";

const { Option } = Select;
class ModelSetting extends React.Component {
  static propTypes = {
    viewer: PropTypes.object.isRequired,
    BIMWINNER: PropTypes.object.isRequired,
    changeSetting: PropTypes.func.isRequired,
    modelSetting: PropTypes.object.isRequired,
    isMobile: PropTypes.bool
  };

  static defaultProps = {
    isMobile: false
  }

  constructor(props) {
    super(props);
    this.inputRefs = React.createRef();
    // this.curModel = '';
    // this.curModelCenter = new THREE.Vector3(0, 0, 0);
    // 防止用户先加载模型后加载UI，导致模型列表为空
    const allModelKeys = this.props.viewer.getViewerImpl().getAllBimModelsKey();
    const m = allModelKeys.map(_key => this.props.viewer.getViewerImpl().getModel(_key));
    const modelKeys = _.keys(m);
    const models = {};
    modelKeys.forEach(modelKey => {
      const name = _.get(m, [modelKey, 'modelConfig', 'modelName'], '');
      if (name) {
        models[modelKey] = name;
      }
    });
    this.state = {
      models,
      curModel: "",
    };
    this.initModelCount = 0;
  }

  shouldComponentUpdate(nextProps) {
    const setting = nextProps.modelSetting;
    if (!_.isEqual(this.props.modelSetting, setting)) {
      const modelKeys = _.keys(setting);
      modelKeys.forEach(key => {
        const xyz = setting[key].basePoint;
        let matrix = this.props.viewer.getViewerImpl().getModelMatrix(key);
        if (matrix) {
          matrix = matrix.clone();
          matrix.elements[12] = xyz[0];
          matrix.elements[13] = xyz[1];
          matrix.elements[14] = xyz[2];
          if (this.initModelCount === modelKeys.length) {
            this.props.viewer.getViewerImpl().setModelMatrix(matrix, key);
          } else {
            this.initModelCount += 1;
          }
          if (key === this.state.curModel) {
            this.inputRefs.current.querySelectorAll("input")
              .forEach((input, index) => {
                input.value = xyz[index];
              });
          }
        }
      });
      this.props.viewer.render();
      return false;
    }
    return true;
  }

  componentDidMount() {
    this.props.viewer.registerModelEventListener(
      this.props.BIMWINNER.BOS3D.EVENTS.ON_LOAD_COMPLETE,
      e => {
        const viewer = this.props.viewer;
        const curModel = viewer.getViewerImpl().getAllBimModelsKey()[0] || "";
        // 初始化所有模型的basePoints
        const allModelKeys = viewer.getViewerImpl().getAllBimModelsKey();
        allModelKeys.forEach(_modelKey => {
          // 获取模型
          const model = viewer.getViewerImpl().getModel(_modelKey);
          const basePoint = model.transform ? model.transform.translation : undefined;
          let xyz = [0, 0, 0];
          if (basePoint) {
            xyz = [basePoint?.x || 0, basePoint?.y || 0, basePoint?.z || 0];
          }
          this.props.changeSetting(_modelKey, "basePoint", xyz);
        });

        // 存储模型的key和名字
        this.setState(state => ({
          models: {
            ...state.models,
            [e.modelKey]: e.target.models[e.modelKey].getConfig().modelName || '',
          },
          curModel,
        }));

        this.changeModel(curModel);
      }
    );
  }

  changeModel(value) {
    const curModel = value;
    const curModelSetting = this.props.modelSetting[curModel];
    if (curModelSetting) {
      const basePoint = curModelSetting.basePoint;
      if (basePoint) {
        this.inputRefs.current.querySelectorAll("input")
          .forEach((input, index) => {
            input.value = basePoint[index];
          });
      }
    }
    this.setState({
      curModel,
    });
    /*
    const matrix = this.props.viewer.viewerImpl.getModelMatrix(this.curModel);
    this.curModelCenter = new THREE.Vector3(
      matrix.elements[12], matrix.elements[13], matrix.elements[14]
    );
     */
  }

  setBasePoint(e) {
    e.preventDefault();
    e.stopPropagation();
    const xyz = [];
    this.inputRefs.current.querySelectorAll("input").forEach(input => {
      if (/^-?[0-9]*(\.[0-9]+)?$/.test(input.value)) {
        xyz.push(Number(input.value));
      }
    });
    if (xyz.length === 3) {
      this.props.changeSetting(this.state.curModel, "basePoint", xyz);
    } else {
      toastr.error("坐标格式错误", "", {
        target: `#${this.props.viewer.viewport}`
      });
    }
  }

  /*
  rotateModel(e) {
    e.preventDefault();
    e.stopPropagation();
    this.props.viewer.viewerImpl.cameraControl.handleRotation(
      Number(e.target.value), 0, this.curModelCenter
    );
    this.props.viewer.render();
  }
  */

  render() {
    const { isMobile } = this.props;
    // console.log(this.curModel, this.state.models, _.keys(this.state.models));
    return (
      <div className={style.settingContainer}>
        <div className={`${style.settingItem} ${style.settingItemModelSelect} `}>
          <div className={style.title}>模型选择</div>
          {
            isMobile && (
              <select
                className={style.select}
                onChange={e => {
                  e.stopPropagation();
                  this.changeModel(e.target.value);
                }}
                value={this.state.curModel}
              >
                {_.keys(this.state.models).map(key => (
                  <option value={key} key={key}>
                    {this.state.models[key]}
                  </option>
                ))}
              </select>
            )
          }
          {
            !isMobile && (
              <div className="boss3d-theme-one-form-form-antd">
                <Select
                  style={{ width: 240, marginTop: 10 }}
                  placeholder="请选择模型选择"
                  onChange={e => { this.changeModel(e) }}
                  dropdownClassName="boss3d-theme-one-form-form-antd-dropdown bos3d-select-dropdown-select-single-has-anticon"
                  getPopupContainer={() => this.props.viewer.viewportDiv}
                  value={this.state.curModel}
                >
                  {
                    _.keys(this.state.models).map(key => (
                      <Option value={key} key={key}>
                        {this.state.models[key]}
                      </Option>
                    ))
                  }
                </Select>
              </div>
            )
          }

        </div>
        <div className={style.basePosition}>
          <div className={style.title}>
            设置模型基点位置：
          </div>
          <div className={style.input} ref={this.inputRefs}>
            <div className={style.basePositionInputItem}>
              <span>X:</span>
              <input type="number" step={1} defaultValue={0} />
            </div>
            <div className={style.basePositionInputItem}>
              <span>Y:</span>
              <input type="number" step={1} defaultValue={0} />
            </div>
            <div className={style.basePositionInputItem}>
              <span>Z:</span>
              <input type="number" step={1} defaultValue={0} />
            </div>
            {
              !isMobile && (
                <button
                  type="button"
                  className="bos-btn bos-btn-primary"
                  onClick={e => { this.setBasePoint(e) }}
                >
                  确定
                </button>
              )
            }
          </div>
        </div>
        {/* <div className={style.settingItem}>
          <span>模型朝向：</span>
          <input
            type="range"
            min={-180}
            max={180}
            step={1}
            defaultValue={0}
            onInput={e => this.rotateModel(e)}
          />
        </div> */}
        {
          isMobile && (
            <div className={style.mobileBtnWrap}>
              <button
                type="button"
                className="bos-btn bos-btn-primary bos-btn-block"
                onClick={e => { this.setBasePoint(e) }}
              >
                确定
              </button>
            </div>

          )
        }
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  viewer: state.system.viewer3D,
  BIMWINNER: state.system.BIMWINNER,
  modelSetting: state.userSetting.modelSetting,
});
const mapDispatchToProps = (dispatch) => ({
  changeSetting: (modelKey, name, value) => {
    dispatch(changeModelSetting(modelKey, name, value));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ModelSetting);
