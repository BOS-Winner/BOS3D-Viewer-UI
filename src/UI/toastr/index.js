/**
 * 提供默认的toastr配置
 * 文档&样例 https://codeseven.github.io/toastr/demo.html
 */
import "toastr/build/toastr.min.css";
import toastr from "toastr";
import "./toastr.less";

toastr.options = {
  "closeButton": false,
  "progressBar": false,
  "positionClass": "toast-top-center",
  "showDuration": 300,
  "hideDuration": 300,
  "timeOut": 3000,
  "extendedTimeOut": 1000,
  "preventDuplicates": true,
  "showEasing": "swing",
  "hideEasing": "linear",
  "showMethod": "fadeIn",
  "hideMethod": "fadeOut",
};

export default toastr;
