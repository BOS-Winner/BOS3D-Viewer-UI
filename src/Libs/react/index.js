import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
