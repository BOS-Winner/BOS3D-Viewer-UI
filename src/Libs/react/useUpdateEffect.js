import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
