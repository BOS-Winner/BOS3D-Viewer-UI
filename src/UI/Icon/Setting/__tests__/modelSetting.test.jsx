import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { genStore } from "jest/initTest";
import ModelSetting from "../ModelSetting";
import toastr from "../../../toastr";

jest.mock('../../../toastr');
let store;
let msContainer;
let msGetByText;
let viewer;

describe("model setting", () => {
  beforeEach(() => {
    store = genStore();
    viewer = store.getState().system.viewer3D;
    const EVENTS = store.getState().system.BIMWINNER.BOS3D.EVENTS;
    viewer.registerModelEventListener = (evt, cb) => {
      if (evt === EVENTS.ON_LOAD_COMPLETE) {
        cb({
          modelKey: 'M1145141919810',
          target: {
            models: {
              M1145141919810: {
                getConfig: () => ({
                  modelName: '佛祖保佑'
                })
              }
            }
          }
        });
        cb({
          modelKey: 'M1145141919811',
          target: {
            models: {
              M1145141919811: {
                getConfig: () => ({
                  modelName: '永无bug'
                })
              }
            }
          }
        });
      }
    };
    const { container, getByText } = render((
      <Provider store={store}>
        <ModelSetting />
      </Provider>
    ));
    msContainer = container;
    msGetByText = getByText;
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("should render model's name list", () => {
    expect(msContainer).toMatchSnapshot();
  });

  it("should change model's base point", () => {
    const setMatrix = jest.fn();
    viewer.getViewerImpl().setModelMatrix = setMatrix;
    const inputs = msContainer.querySelectorAll("input");
    inputs[0].value = '1000';
    inputs[1].value = '-1000';
    inputs[2].value = '1000';
    const button = msGetByText('确定');
    button.click();
    expect(setMatrix).toHaveBeenCalledTimes(1);
    expect(setMatrix.mock.calls[0]).toMatchSnapshot();

    const select = msGetByText('模型选择：').nextElementSibling;
    fireEvent.change(select, {
      target: {
        value: 'M1145141919811'
      }
    });
    inputs[0].value = '-2000';
    inputs[1].value = '2000';
    inputs[2].value = '-2000';
    button.click();
    expect(setMatrix).toHaveBeenCalledTimes(3);
    expect(setMatrix.mock.calls[1]).toMatchSnapshot();
    expect(setMatrix.mock.calls[2]).toMatchSnapshot();

    fireEvent.change(select, {
      target: {
        value: 'M1145141919810'
      }
    });
    expect(inputs[0].value).toBe('1000');
    expect(inputs[1].value).toBe('-1000');
    expect(inputs[2].value).toBe('1000');
  });

  it("should cause error when using error input", () => {
    const inputs = msContainer.querySelectorAll("input");
    inputs[0].value = 'fxxk';
    const button = msGetByText('确定');
    button.click();
    expect(toastr.error).toBeCalledTimes(1);
  });

  it("shouldn't update matrix when model not found", () => {
    const setMatrix = jest.fn();
    viewer.getViewerImpl().setModelMatrix = setMatrix;
    const select = msGetByText('模型选择：').nextElementSibling;
    fireEvent.change(select, {
      target: {
        value: 'Model not found'
      }
    });
    const button = msGetByText('确定');
    button.click();
    expect(setMatrix).toBeCalledTimes(0);
  });
});
