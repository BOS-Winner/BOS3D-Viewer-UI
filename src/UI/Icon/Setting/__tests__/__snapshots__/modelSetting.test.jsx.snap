// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model setting should change model's base point 1`] = `
Array [
  Matrix4 {
    "elements": Array [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      1000,
      -1000,
      1000,
      1,
    ],
  },
  "M1145141919810",
]
`;

exports[`model setting should change model's base point 2`] = `
Array [
  Matrix4 {
    "elements": Array [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      1000,
      -1000,
      1000,
      1,
    ],
  },
  "M1145141919810",
]
`;

exports[`model setting should change model's base point 3`] = `
Array [
  Matrix4 {
    "elements": Array [
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      -2000,
      2000,
      -2000,
      1,
    ],
  },
  "M1145141919811",
]
`;

exports[`model setting should render model's name list 1`] = `
<div>
  <div
    class="settingContainer"
  >
    <div
      class="settingItem"
    >
      <span>
        模型选择：
      </span>
      <select>
        <option
          value="M1145141919810"
        >
          佛祖保佑
        </option>
        <option
          value="M1145141919811"
        >
          永无bug
        </option>
      </select>
    </div>
    <div
      class="basePosition"
    >
      <div>
        设置模型基点位置：
      </div>
      <div
        class="input"
      >
        <span>
          X:
        </span>
        <input
          step="1"
          type="number"
          value="0"
        />
        <span>
          Y:
        </span>
        <input
          step="1"
          type="number"
          value="0"
        />
        <span>
          Z:
        </span>
        <input
          step="1"
          type="number"
          value="0"
        />
        <button
          class="button"
          type="button"
        >
          确定
        </button>
      </div>
    </div>
  </div>
</div>
`;
